### ISSUE TITLE
[Wiki]: Document API response structures

### ISSUE THREAD

**Issue by jsvana** [jjjonesjr33/petlibro#14]
### Device model name

N/A

### Device model number

N/A

### Device type

Other

### Feature name

API documentation

### Describe the feature

I believe it would be helpful to developers and maintainers to have the PETLIBRO API response structures documented in the plugin's wiki. We can use that to add structure to the code and iterate quicker on new change requests.

### What problem does this feature solve?

Developer velocity

### Available to help with the feature development

- [X] I agree to help with testing or providing relevant information for this feature request

**Comment by jjjonesjr33** [comment]
We should be able to do this, I just need to capture what outputs I am getting per device.

**Comment by jjjonesjr33** [comment]
Linking documentation from feature/pull requests https://github.com/jjjonesjr33/petlibro/issues/20, https://github.com/jjjonesjr33/petlibro/issues/24, https://github.com/jjjonesjr33/petlibro/issues/25, and https://github.com/jjjonesjr33/petlibro/pull/23

Will add to the wiki once we have more info from everything.

**Comment by jjjonesjr33** [comment]
Linking documentation from feature/pull requests https://github.com/jjjonesjr33/petlibro/issues/26 and https://github.com/jjjonesjr33/petlibro/issues/27

**Comment by jjjonesjr33** [comment]
Linking documentation from feature/pull requests https://github.com/jjjonesjr33/petlibro/issues/16

**Comment by Nistp** [comment]
@jjjonesjr33 I've been wondering for a bit, but how are you getting the data? Are you using a proxy? Running an Android emulator? Do you mind sharing so I could also help?

**Comment by C4-Dimitri** [comment]
> [@jjjonesjr33](https://github.com/jjjonesjr33) I've been wondering for a bit, but how are you getting the data? Are you using a proxy? Running an Android emulator? Do you mind sharing so I could also help?

I'll respond as i have been getting a lot of this data.

I am using an android emulator, and i am using an application called http toolkit. It walks you through how to then install the CA Cert on your emulator, you also need to enable wireless adb debugging in the developer settings on android. When opening the petlibro app open it first without http toolkit attached as on login it seems to do some verification that fails if http toolkit is running. Hook in http toolkit by using adb the card for this should show automatically in http toolkit if youve done everything right. Click the card and you are good to go. You can now see the requests made from your android device.

Http toolkit is free, but they have a paid version that can parse some more complex data (not needed for our purposes) and you can also resend post/get requests which can be usefull so you dont have to test EVERYTHING in HA.

**Comment by Nistp** [comment]
@C4-Dimitri, thank you very much for describing your workflow, quite helpful. Could  you share the emulator that you are using?

**Comment by C4-Dimitri** [comment]
> [@C4-Dimitri](https://github.com/C4-Dimitri), thank you very much for describing your workflow, quite helpful. Could you share the emulator that you are using?

Just using android studio :)

**Comment by christianprescott** [comment]
Hi, is there any interest in turning the API documentation into something formal like an OpenAPI specification? @jjjonesjr33 

I've taken a crack at it [on this branch.](https://github.com/christianprescott/petlibro/compare/dev...christianprescott:petlibro:openapi-docs#diff-b1b21eca4df802bb204cc7c8d8d5f2c628ee8214dffa918252133859b44a3784R354) The spec is a YAML format file that is verbose and not particularly pleasant to look at, but it is machine readable and enables other tools to use this info. For example, [this page](https://christianprescott.com/petlibro/docs/api/petlibro/) reads the spec and turns it into human readable and _interactive_ API docs.

This would replace the wiki page and would be the place to add request and response data discovered during development [like this](https://github.com/jjjonesjr33/petlibro/issues/96#issuecomment-3134112490). In the future it could be used to test the python code in this project with simulated requests, or to detect changes in Petlibro's API when the responses vary.

I'm commenting here because getting it into this shape and into version control would meet the goal of this issue, even though I don't have the complete set of APIs documented. (Ultimately, my motivation is I'd like to use the Petlibro API for other non-Home Assistant purposes and this project seems to have the most mature implementation of it!)

**Comment by C4-Dimitri** [comment]
@christianprescott, I think that this is a good idea, but i dont see the reason for it replacing the wiki entirely. Why not just have both? A more human readable version, and a more machine readable version (or think beginner vs advanced developer).

**Comment by christianprescott** [comment]
The reason is because it would be convenient to have a "single source of truth." When there are two versions it is possible for one to fall behind, or for them to have different instructions which could be confusing. And it requires someone to do twice as much work when they want to add an API since they have to go to two places. Because the OpenAPI has more specific detail and utility that would be my preference.

But I can see your point about how wiki might be easier for a beginner to read! I don't feel strongly about this so I wouldn't demand that wiki be removed.

**Comment by C4-Dimitri** [comment]
At the moment I think the repo is currently more geared towards being additive vs subtractive.

**Comment by christianprescott** [comment]
Yep, understood. I used this OpenAPI format on another petlibro project and it significantly improved my velocity. I used the API [docs web page](https://christianprescott.com/petlibro/docs/api/petlibro/) for testing, then used its generated API client to automatically create code in my language of choice:

```js
const devicesApi = new petlibro.DevicesApi(config);
const { data: [device] } = await devicesApi.listDevices();
const feedResponse = await devicesApi.manualFeed({
  deviceSn: device.deviceSn,
  grainNum: 1,
  requestId: "a1b2c3d4e5f6a1b2c3d4e5f6a1b2c3d4",
});
```

Let me know if you're ever interested in pursuing this. I don't have a ton of petlibro devices to sample from but can assist with setting up the tools.
